sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSchemaInferenceSuite.scala

```diff
@@ -0,0 +1,352 @@
+    // and that the raw table returned by the Hive client doesn't have any Spark SQL properties
+    // set (table needs to be obtained from client since HiveExternalCatalog filters these
+    // properties out).
+    assert(!externalCatalog.getTable(DATABASE, tableName).schemaPreservesCase)
+    val rawTable = client.getTable(DATABASE, tableName)
+    assert(rawTable.properties.filterKeys(_.startsWith(SPARK_SQL_PREFIX)) == Map.empty)
+    schema
+  }
+
+  private def withTestTables(
+    fileType: String)(f: (String, Seq[String], Seq[String], StructType) => Unit): Unit = {
+    // Test both a partitioned and unpartitioned Hive table
+    val tableFields = Seq(
+      (Seq("fieldOne"), Seq("partCol1", "partCol2")),
+      (Seq("fieldOne", "fieldTwo"), Seq.empty[String]))
+
+    tableFields.foreach { case (fields, partCols) =>
+      withTempDir { dir =>
+        val tableName = "test_table"
+        val schema = setupExternalTable(tableName, fileType, fields, partCols, dir)
+        try f(
+          tableName,
+          fields,
+          partCols,
+          schema) finally spark.sql(s"DROP TABLE IF EXISTS $tableName")
+      }
+    }
+  }
+
+  private def withFileTypes(f: (String) => Unit): Unit
+    = Seq(ORC_FILE_TYPE, PARQUET_FILE_TYPE).foreach(f)
+
+  private def withInferenceMode(mode: InferenceMode)(f: => Unit): Unit = {
+    withSQLConf(
+      HiveUtils.CONVERT_METASTORE_ORC.key -> "true",
+      SQLConf.HIVE_CASE_SENSITIVE_INFERENCE.key -> mode.toString)(f)
+  }
+
+  private val inferenceKey = SQLConf.HIVE_CASE_SENSITIVE_INFERENCE.key
+
+  private def testFieldQuery(tableName: String, fields: Seq[String]): Unit = {
+    if (!fields.isEmpty) {
+      val query = s"SELECT * FROM $tableName WHERE ${Random.shuffle(fields).head} >= 0"
+      assert(spark.sql(query).count == NUM_RECORDS)
+    }
+  }
+
+  private def testSchemaQuery(tableName: String, expectedSchema: StructType): Unit
+    = assert(spark.sql(s"SELECT * FROM $tableName").schema == expectedSchema)
+
+  withFileTypes { fileType =>
+    test(s"$fileType: schema should be inferred and saved when INFER_AND_SAVE is specified") {
+      withInferenceMode(INFER_AND_SAVE) {
+        withTestTables(fileType) { (tableName, fields, partCols, schema) =>
+          testFieldQuery(tableName, fields)
+          testFieldQuery(tableName, partCols)
+          // Test that the case-sensitive schema was storied as a table property after inference
+          testSchemaQuery(tableName, schema)
+
+          // Verify the catalog table now contains the updated schema and properties
+          val catalogTable = externalCatalog.getTable(DATABASE, tableName)
+          assert(catalogTable.schemaPreservesCase)
+          assert(catalogTable.schema == schema)
+          assert(catalogTable.partitionColumnNames == partCols.map(_.toLowerCase))
+          val rawTable = client.getTable(DATABASE, tableName)
+          assert(rawTable.properties.contains(HiveExternalCatalog.DATASOURCE_SCHEMA_NUMPARTS))
```

Comment:
as long as `externalCatalog.getTable` returns corrected table schema, we don't care about the special table properties. Let's not test this.